### DISCUSSION TITLE
Nested forks & joins

### DISCUSSION THREAD

**Discussion by tonymathewt** [elsa-workflows/elsa-core#2746] [Q&A] [answered]
I am trying to have nested forks/joins. For the below scenario:

The workflow when triggered forks into two children (ChildOne & ChildTwo). Both these children could have a Success or a Failed signal which are sub-forks with branches under each. I need to wait for either of these two signals from both the children to complete the workflow. The programmatic workflow I built does not do this. It completes as soon as it finishes any one signal. Moreover its not firing the HttpEndponit I have written at the end during completion. Looks like something is wrong with the way I wrote. I want this built programmatically since there is a possibility of some dynamic nature in the future.

I am attaching the zipped project here which has the workflow named 'ProvisioningWorkflow'. Could anyone take a look at it please. Have also attached the postman collection with endpoints to trigger the flow, send childOne failed and send ChildTwo success signals. The workflowinstance Id being set in the variable needs to be updated for each instance. 
[ELSASignals.zip](https://github.com/elsa-workflows/elsa-core/files/8041172/ELSASignals.zip)
[TestElsaForking.postman_collection.zip](https://github.com/elsa-workflows/elsa-core/files/8041179/TestElsaForking.postman_collection.zip)

**Comment by sfmskywalker** [comment, marked as answer]
Hi Tony,

As discussed on Discord:

One issue is that the connections between the nested joins and the final join needs to be corrected:

![image](https://user-images.githubusercontent.com/938393/154545347-7a90470a-546b-4e93-9055-1ba44d2e1b78.png)

To fix that, change this:

![image](https://user-images.githubusercontent.com/938393/154545452-d9746537-d151-404a-8503-353045e9da13.png)

To this:

![image](https://user-images.githubusercontent.com/938393/154545529-2ab87587-4782-4f8d-8088-7ea2c3e806e6.png)

Which will result in a correctly-joined diagram:

![image](https://user-images.githubusercontent.com/938393/154545597-7c80883f-3552-4e0d-8e3f-0211584bf55f.png)

But that was only part of the issue. As it turns out, there was a bug in the **Join** activity logic, where it was removing blocking activities too eagerly.

I pushed a changed that fixes this and added a (simplified) [sample](https://github.com/elsa-workflows/elsa-core/tree/master/src/samples/aspnet/Elsa.Samples.NestedForks) based on your code 😄 

Thanks for reporting and for the great repro, that helped a lot :)
